datatransfer/README.md

```diff
@@ -0,0 +1,156 @@
+    }
+    ```
+
+1. If needed, build out your voucher struct and its validator. 
+    
+    A push or pull request must include a voucher. The voucher's type must have been registered with 
+    the node receiving the request before it's sent, otherwise the request will be rejected.  
+
+    [datatransfer.Voucher](https://github.com/filecoin-project/go-fil-components/blob/21dd66ba370176224114b13030ee68cb785fadb2/datatransfer/types.go#L17)
+    and [datatransfer.Validator](https://github.com/filecoin-project/go-fil-components/blob/21dd66ba370176224114b13030ee68cb785fadb2/datatransfer/types.go#L153)
+    are the interfaces used for validation of graphsync datatransfer messages.  Voucher types plus a Validator for them must be registered
+    with the peer to whom requests will be sent.  
+
+#### Example Toy Voucher and Validator
+```go
+type myVoucher struct {
+	data string
+}
+
+func (v *myVoucher) ToBytes() ([]byte, error) {
+	return []byte(v.data), nil
+}
+
+func (v *myVoucher) FromBytes(data []byte) error {
+	v.data = string(data)
+	return nil
+}
+
+func (v *myVoucher) Type() string {
+	return "FakeDTType"
+}
+
+type myValidator struct {
+	ctx                 context.Context
+	validationsReceived chan receivedValidation
+}
+
+func (vl *myValidator) ValidatePush(
+	sender peer.ID,
+	voucher datatransfer.Voucher,
+	baseCid cid.Cid,
+	selector ipld.Node) error {
+    
+    v := voucher.(*myVoucher)
+    if v.data == "" || v.data != "validpush" {
+        return errors.New("invalid")
+    }   
+
+	return nil
+}
+
+func (vl *myValidator) ValidatePull(
+	receiver peer.ID,
+	voucher datatransfer.Voucher,
+	baseCid cid.Cid,
+	selector ipld.Node) error {
+
+    v := voucher.(*myVoucher)
+    if v.data == "" || v.data != "validpull" {
+        return errors.New("invalid")
+    }   
+
+	return nil
+}
+
+```
+
+
+Please see 
+[go-fil-components/blob/master/datatransfer/types.go](https://github.com/filecoin-project/go-fil-components/blob/master/datatransfer/types.go) 
+for more detail.
+
+
+### Register a validator
+Before sending push or pull requests, you must register a `datatransfer.Voucher` 
+by its `reflect.Type` and `dataTransfer.RequestValidator` for vouchers that
+must be sent with the request.  Using the trivial examples above:
+```go
+    func NewGraphsyncDatatransfer(h host.Host, gs graphsync.GraphExchange) {
+        dt := datatransfer.NewGraphSyncDataTransfer(h, gs)
+        vouch := &myVoucher{}
+        mv := &myValidator{} 
+        dt.RegisterVoucherType(reflect.TypeOf(&vouch), &mv)
+    }
+```
+    
+For more detail, please see the [unit tests](https://github.com/filecoin-project/go-fil-components/blob/master/datatransfer/impl/graphsync/graphsync_impl_test.go).
+
+### Open a Push or Pull Request
+For a push or pull request you need a context, a `datatransfer.Voucher`, a host `peer.ID`, a base `cid.CID`
+and a selector.
+```go
+    func NewGraphsyncDatatransfer(ctx context.Context, h host.Host, gs graphsync.GraphExchange) {
```

Comment:
probably get rid the containing code (assume they read the instructions above), just list call to push/pull